go/vt/tabletserver/gorpcqueryservice/sqlquery.go

```diff
@@ -47,32 +91,57 @@ func (sq *SqlQuery) Rollback(ctx context.Context, session *proto.Session, noOutp
 // Execute is exposing tabletserver.SqlQuery.Execute
 func (sq *SqlQuery) Execute(ctx context.Context, query *proto.Query, reply *mproto.QueryResult) (err error) {
 	defer sq.server.HandlePanic(&err)
-	execErr := sq.server.Execute(callinfo.RPCWrapCallInfo(ctx), query, reply)
-	tabletserver.AddTabletErrorToQueryResult(execErr, reply)
+	tErr := sq.server.Execute(callinfo.RPCWrapCallInfo(ctx), query, reply)
+	tabletserver.AddTabletErrorToQueryResult(tErr, reply)
 	if *tabletserver.RPCErrorOnlyInReply {
 		return nil
 	}
-	return execErr
+	return tErr
 }
 
 // StreamExecute is exposing tabletserver.SqlQuery.StreamExecute
 func (sq *SqlQuery) StreamExecute(ctx context.Context, query *proto.Query, sendReply func(reply interface{}) error) (err error) {
 	defer sq.server.HandlePanic(&err)
-	return sq.server.StreamExecute(callinfo.RPCWrapCallInfo(ctx), query, func(reply *mproto.QueryResult) error {
+	tErr := sq.server.StreamExecute(callinfo.RPCWrapCallInfo(ctx), query, func(reply *mproto.QueryResult) error {
 		return sendReply(reply)
 	})
+	if tErr == nil {
+		return nil
+	}
+	if *tabletserver.RPCErrorOnlyInReply {
+		// If there was an app error, send a QueryResult back with it.
+		qr := new(mproto.QueryResult)
+		tabletserver.AddTabletErrorToQueryResult(tErr, qr)
```

Comment:
Unfortunately, there isn't any way to attach an error to an existing message. Each message is sent as the DB streams the results back: https://github.com/youtube/vitess/blob/d4a89685afa4f3b60af38805e0f3d9392478d539/go/vt/dbconnpool/connection.go#L84

If a row is successfully received, it is sent. If an error is received, no more messages are sent, and an error is returned instead.

Basically, by the time an error happens, every message that was going to be sent has already been sent. That's why we need to send one last message (as I'm doing here) to send the error.

I think the alternative way to do it would be to use the API that Alain has for his proto3 stuff - return a StreamExecuteResponse that would contain a QueryResult and an Err, and clients would unwrap that to figure out what to do. In practice, that would look pretty much the same as what's happening here, but backwards compatibility would be maintained by adding an new end-point.